### EDA STEP 1: UNDERSTANDING THE PROBLEM AND HYPOTESİS

#### We have 2 different datas: fighter stat and fight datas. Moreover, so many people think that physical features like reach, height or age can be very tremendously important for winning a fight. Therefore, my goal is trying to prove that reach, height and age really important for winning an ufc fight so you can think that my hypotesis is that physical attributes like reach, height and age are really important for winning or losing a ufc fight.

### Importing Libraries

In [101]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### creating dataframes

In [102]:
fighters_df = pd.read_csv("../datasets/fighter_stats.csv") # fighter stats csv to dataframe
fights_df = pd.read_csv("../datasets/raw_fights.csv") # fight datas csv to dataframe

###### it is good to be sure that we get datas correctly color 

In [103]:
fighters_df.head(5)

name  wins  losses  height  weight   reach    stance   age  \
0      Amanda Ribas  12.0     5.0  160.02   56.70  167.64  Orthodox  30.0   
1    Rose Namajunas  13.0     6.0  165.10   56.70  165.10  Orthodox  31.0   
2     Karl Williams  10.0     1.0  190.50  106.59  200.66  Orthodox  34.0   
3       Justin Tafa   7.0     4.0  182.88  119.75  187.96  Southpaw  30.0   
4  Edmen Shahbazyan  13.0     4.0  187.96   83.91  190.50  Orthodox  26.0   

   SLpM  sig_str_acc  SApM  str_def  td_avg  td_acc  td_def  sub_avg  
0  4.63         0.40  3.40     0.61    2.07    0.51    0.85      0.7  
1  3.69         0.41  3.51     0.63    1.38    0.47    0.59      0.5  
2  2.87         0.52  1.70     0.60    4.75    0.50    1.00      0.2  
3  4.09         0.54  5.02     0.47    0.00    0.00    0.50      0.0  
4  3.60         0.52  4.09     0.45    2.24    0.38    0.63      0.6

In [114]:
fights_df.head(5)

Fight_Id Win/No Contest/Draw         Fighter_1  \
0  d61373d9fdf01a12                 win     Brendan Allen   
1  cdd565a16975c711                 win       Mike Malott   
2  857d274e47aa8c4e                 win    Aiemann Zahabi   
3  c82acd7b17915591                 win      Manon Fiorot   
4  a18edcb638ba7e39                 win  Charles Jourdain   

              Fighter_2  KD_1  KD_2  STR_1  STR_2  TD_1  TD_2  SUB_1  SUB_2  \
0     Reinier de Ridder   0.0   0.0   22.0    9.0   1.0   3.0    1.0    1.0   
1         Kevin Holland   0.0   0.0   49.0   52.0   1.0   1.0    1.0    0.0   
2           Marlon Vera   0.0   1.0   81.0   56.0   0.0   0.0    0.0    0.0   
3  Jasmine Jasudavicius   0.0   0.0   16.0    2.0   0.0   0.0    0.0    0.0   
4           Davey Grant   1.0   0.0   25.0    9.0   0.0   0.0    1.0    0.0   

        Weight_Class                Method  Round  Fight_Time  \
0       Middleweight                KO/TKO      4         300   
1       Welterweight                 U-DEC      3         300   
2       Bantamweight                 S-DEC      3         300   
3  Women's Flyweight        KO/TKO Punches      1          74   
4       Bantamweight  SUB Guillotine Choke      1         185   

           Event_Id  
0  38e5d9dcb0fddc42  
1  38e5d9dcb0fddc42  
2  38e5d9dcb0fddc42  
3  38e5d9dcb0fddc42  
4  38e5d9dcb0fddc42

In [117]:
fights_df.shape

(8402, 17)

##### Now we can start to analyze our data. Let's start with fights_df

###### There are 17 columns and 8402 rows. 
###### Let's explanining columns. 
###### Fight_Id: id of the match.
###### Win/No Contest/Draw: how the match ended
###### Fighter_1: first fighter
###### Fİghther_2: second fighter
###### KD_1: Number of knockdowns landed by first fighter
###### KD_2: Number of knockdowns landed by second fighter
###### STR_1: Number of Significant Strikes landed by the first fighter
###### STR_2: Number of Significant Strikes landed by the first fighter
###### TD_1: Takedowns landed by the first fighter
###### TD_2: Takedowns landed by the second fighter
###### SUB_1: Number of submission attempts by first fighter 
###### SUB_2: Number of submission attempts by second fighter
###### Weight_Class: The weight division of the fight
###### Method: How the fight ended (TKO, KO, Decision...)
###### Round: The round which the fight ended
###### Fight_Time: Time elapsed when the fight concluded
###### Event_Id: Id of the event which fight happend


In [105]:
fights_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8402 entries, 0 to 8401
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Fight_Id             8402 non-null   object
 1   Win/No Contest/Draw  8402 non-null   object
 2   Fighter_1            8402 non-null   object
 3   Fighter_2            8402 non-null   object
 4   KD_1                 8402 non-null   object
 5   KD_2                 8402 non-null   object
 6   STR_1                8402 non-null   object
 7   STR_2                8402 non-null   object
 8   TD_1                 8402 non-null   object
 9   TD_2                 8402 non-null   object
 10  SUB_1                8402 non-null   object
 11  SUB_2                8402 non-null   object
 12  Weight_Class         8402 non-null   object
 13  Method               8402 non-null   object
 14  Round                8402 non-null   int64 
 15  Fight_Time           8402 non-null   object
 16  Event_

##### Since some columns which has to be integer (like STR_1 STR_2 etc.) are object I have to change their dtype to numeric values so that I can analyze them in an easier way.

In [106]:
fights_df.describe()

Round
count  8402.000000
mean      2.353368
std       1.017514
min       1.000000
25%       1.000000
50%       3.000000
75%       3.000000
max       5.000000

In [107]:
fights_df.dtypes

Fight_Id               object
Win/No Contest/Draw    object
Fighter_1              object
Fighter_2              object
KD_1                   object
KD_2                   object
STR_1                  object
STR_2                  object
TD_1                   object
TD_2                   object
SUB_1                  object
SUB_2                  object
Weight_Class           object
Method                 object
Round                   int64
Fight_Time             object
Event_Id               object
dtype: object

In [108]:
fights_df["Fight_Time"] = fights_df["Fight_Time"].apply(lambda x: int(x.split(":")[0]) * 60 + int(x.split(":")[1])) # thx to this we can turn object type to integer and change minute:second to seconds
integer_columns = ["KD_1", "KD_2", "STR_1", "STR_2", "TD_1", "TD_2", "SUB_1", "SUB_2"]
fights_df[integer_columns] = fights_df[integer_columns].apply(pd.to_numeric, errors="coerce")

In [109]:
fights_df["Fight_Time"]

0       300
1       300
2       300
3        74
4       185
       ... 
8397    170
8398    292
8399    733
8400     58
8401     20
Name: Fight_Time, Length: 8402, dtype: int64

In [110]:
fights_df.dtypes

Fight_Id                object
Win/No Contest/Draw     object
Fighter_1               object
Fighter_2               object
KD_1                   float64
KD_2                   float64
STR_1                  float64
STR_2                  float64
TD_1                   float64
TD_2                   float64
SUB_1                  float64
SUB_2                  float64
Weight_Class            object
Method                  object
Round                    int64
Fight_Time               int64
Event_Id                object
dtype: object

##### Since dtypes are changed we can move to our other dataframe, fighters_df.

In [111]:
fighters_df.head(5)

name  wins  losses  height  weight   reach    stance   age  \
0      Amanda Ribas  12.0     5.0  160.02   56.70  167.64  Orthodox  30.0   
1    Rose Namajunas  13.0     6.0  165.10   56.70  165.10  Orthodox  31.0   
2     Karl Williams  10.0     1.0  190.50  106.59  200.66  Orthodox  34.0   
3       Justin Tafa   7.0     4.0  182.88  119.75  187.96  Southpaw  30.0   
4  Edmen Shahbazyan  13.0     4.0  187.96   83.91  190.50  Orthodox  26.0   

   SLpM  sig_str_acc  SApM  str_def  td_avg  td_acc  td_def  sub_avg  
0  4.63         0.40  3.40     0.61    2.07    0.51    0.85      0.7  
1  3.69         0.41  3.51     0.63    1.38    0.47    0.59      0.5  
2  2.87         0.52  1.70     0.60    4.75    0.50    1.00      0.2  
3  4.09         0.54  5.02     0.47    0.00    0.00    0.50      0.0  
4  3.60         0.52  4.09     0.45    2.24    0.38    0.63      0.6

In [112]:
fighters_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2479 entries, 0 to 2478
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         2478 non-null   object 
 1   wins         2478 non-null   float64
 2   losses       2478 non-null   float64
 3   height       2478 non-null   float64
 4   weight       2478 non-null   float64
 5   reach        1823 non-null   float64
 6   stance       2401 non-null   object 
 7   age          2318 non-null   float64
 8   SLpM         2478 non-null   float64
 9   sig_str_acc  2478 non-null   float64
 10  SApM         2478 non-null   float64
 11  str_def      2478 non-null   float64
 12  td_avg       2478 non-null   float64
 13  td_acc       2478 non-null   float64
 14  td_def       2478 non-null   float64
 15  sub_avg      2478 non-null   float64
dtypes: float64(14), object(2)
memory usage: 310.0+ KB


In [113]:
fighters_df.describe()

wins       losses       height       weight        reach  \
count  2478.000000  2478.000000  2478.000000  2478.000000  1823.000000   
mean     14.399112     6.111380   178.384262    76.876852   182.035656   
std       9.853474     4.548011     8.851777    17.976646    10.654129   
min       0.000000     0.000000   152.400000    52.160000   147.320000   
25%       9.000000     3.000000   172.720000    65.770000   175.260000   
50%      13.000000     5.000000   177.800000    77.110000   182.880000   
75%      18.000000     8.000000   185.420000    83.910000   190.500000   
max     253.000000    53.000000   210.820000   349.270000   213.360000   

               age         SLpM  sig_str_acc         SApM      str_def  \
count  2318.000000  2478.000000  2478.000000  2478.000000  2478.000000   
mean     37.727783     2.932623     0.406029     3.388567     0.488910   
std       7.251188     1.736693     0.151297     2.027258     0.162305   
min      19.000000     0.000000     0.000000     0.000000     0.000000   
25%      32.000000     1.810000     0.360000     2.320000     0.450000   
50%      37.000000     2.900000     0.430000     3.230000     0.520000   
75%      42.000000     3.920000     0.500000     4.290000     0.580000   
max      81.000000    23.330000     1.000000    42.000000     1.000000   

            td_avg       td_acc       td_def      sub_avg  
count  2478.000000  2478.000000  2478.000000  2478.000000  
mean      1.393543     0.322571     0.494972     0.602179  
std       1.512821     0.256928     0.297551     1.057863  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.120000     0.070000     0.300000     0.000000  
50%       1.000000     0.330000     0.550000     0.300000  
75%       2.065000     0.480000     0.710000     0.800000  
max      13.950000     1.000000     1.000000    16.400000

In [118]:
fighters_df.shape

(2479, 16)

In [119]:
fighters_df.head(5)

name  wins  losses  height  weight   reach    stance   age  \
0      Amanda Ribas  12.0     5.0  160.02   56.70  167.64  Orthodox  30.0   
1    Rose Namajunas  13.0     6.0  165.10   56.70  165.10  Orthodox  31.0   
2     Karl Williams  10.0     1.0  190.50  106.59  200.66  Orthodox  34.0   
3       Justin Tafa   7.0     4.0  182.88  119.75  187.96  Southpaw  30.0   
4  Edmen Shahbazyan  13.0     4.0  187.96   83.91  190.50  Orthodox  26.0   

   SLpM  sig_str_acc  SApM  str_def  td_avg  td_acc  td_def  sub_avg  
0  4.63         0.40  3.40     0.61    2.07    0.51    0.85      0.7  
1  3.69         0.41  3.51     0.63    1.38    0.47    0.59      0.5  
2  2.87         0.52  1.70     0.60    4.75    0.50    1.00      0.2  
3  4.09         0.54  5.02     0.47    0.00    0.00    0.50      0.0  
4  3.60         0.52  4.09     0.45    2.24    0.38    0.63      0.6

##### Since this dtypes are float for columns; wins, losses, height weight reach etc. we do not need to turn this data types to
##### Let's analyze columns for fighters_df

###### There are 16 columns and 2479 rows
###### name: name of the fighter
###### wins: total wins of that fighter
###### losses: total losses of that fighter
###### height: height of that fighter
###### reach: reach of that fighter
###### stance: stance of that fighter (Orthodox: traditional left hand forward fighting stance and Southpaw: right hand forward fighting stance)
###### age: age of that fighter
###### SLpM: Significant strikes landed per minute by that fighter
###### sig_str_Acc: Significant Strikes Accuracy (ratio of landed to attempted strikes)
###### SApM: Significant Strikes Absorbed per minute by that fighter
###### str_def: Strike Defence ratio (ratio of strikes successfully avoided)
###### td_avg: Takedowns landed by match
###### td_acc: Takedowns accuracy